### ISSUE TITLE
Configuring AWS clients

### ISSUE THREAD

**Issue by sagikazarmark** [temporalio/temporal-aws-sdk#2]
Currently activities receive session and config directly. This serves well with the following two use cases:

- User has a single AWS account to operate on.
- User has multiple AWS credentials, but can run multiple workers and separate tasks using task queues.

There is a third use case, however, which we (Banzai Cloud) are heavy users of: there is an unbound number of AWS credentials that we have to use and need to start new sessions in every activity.

The first two use cases are probably the most common ones, but I don't think the latter is without precedent either.

Therefore I suggest introducing an abstraction layer for instantiating sessions (and in turn, AWS clients).

There is a problem in the third use case that we need to solve first. In our scenario, there is a "secret ID" that identifies the AWS credential stored in a secret store. We need to pass that down to activities. Currently (in our internal code base) we make that secret ID part of the activity input, but that would probably be not really nice for a generic purpose library (even if a `SecretID` parameter would be generic enough for alternate use cases).

Instead, we can use the context for this purpose. With ContextPropagators, we can inject a secret ID into the context and retrieve it in activities. Although the secret ID is not really a "metadata" (which is what these context propagation solutions are primarily for), it's a good enough solution for now.

The last decision we need to make is whether we want to return `Session` or `Config` objects from this abstraction.
Since AWS recommends [caching sessions](https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/sessions.html) for performance reasons, returning `Session` objects would support the first two use cases better.

Based on the above, we can now establish an interface for configuring AWS clients:

```go
// SessionFactory returns an aws.Session based on the activity context.
type SessionFactory interface{
	Session(ctx context.Context) (*aws.Session, error)
}
```

A default implementation of the above factory could return a single session for all requests:

```go
type cachedSessionFactory struct{
	session *aws.Session
}

func (s *cachedSessionFactory) Session(ctx workflow.Context) (*aws.Session, error) {
	return s.session, nil
}
```

Clients then can accept a `SessionFactory` instead of a session and a configuration and call it for each activity:

```
type SomeActivities struct {
	sessionFactory SessionFactory
}

func NewSomeActivities(sessionFactory SessionFactory) *SomeActivities {
	return &SomeActivities{sessionFactory: sessionFactory}
}

func (a *SomeActivities) SomeOperation(ctx context.Context, input *some.SomeOperationInput) (*some.SomeOperationOutput, error) {
	sess, err := a.sessionFactory.Session(ctx)
	if err != nil {
		return nil, err
	}

	client := some.New(sess)

	return a.client.AddTagsToCertificateWithContext(ctx, input)
}
```

Although it requires a bit more generated code, the overhead of creating a new client should be negligible.

Obviously, the above change adds some error handling to the code, but in case of the cached session it will always be nil, in case of an alternative implementation (eg. accessing a secret store) retries should be done by the implementation, thus we can safely assume that the error here is fatal. We should probably try to encode it anyway (see #1).

An alternative solution would be going for client factories instead (again, with a cached implementation that spares the cost of creating the client in each activity), but that would mean generating much more code and putting an overhead on the instantiation side.

**Comment by mfateev** [comment]
I don't like the idea of recreating client on every request. It might be required for the multisession approach, but looks too heavyweight for the global session one. I would try to come up with a way that behaves differently based on the scenario. 

In the worst case, we can generate two different activity implementations for these cases.

**Comment by sagikazarmark** [comment]
Initializing the client is actually not heavy at all. It just creates a new struct. It doesn't do anything else.

Creating the session is much heavier, because it reads all kinds of stuff from the environment.

As an alternative, activities could support both:

- if there is a client set, use it
- if there is no client, but there is a session factory: create a new client
- if there is no client, nor a session factory: this should never happen

WDYT?

**Comment by sagikazarmark** [comment]
Here is an alternative implementation:

```go
type SomeActivities struct {
	client         someiface.SomeAPI
	sessionFactory SessionFactory
}

func NewSomeActivities(session *session.Session, config ...*aws.Config) *SomeActivities {
	client := some.New(session, config...)
	return &SomeActivities{client: client}
}

func NewSomeActivitiesWithSessionFactory(sessionFactory SessionFactory) *SomeActivities {
	return &SomeActivities{sessionFactory: sessionFactory}
}

func (a *SomeActivities) getClient(ctx context.Context) (someiface.SomeAPI, error) {
	if a.client != nil { // No need to protect with mutex: we know the client never changes
		return a.client 
	}

	sess, err := a.sessionFactory.Session(ctx)
	if err != nil {
		return nil, err
	}

	return some.New(sess), err
}

func (a *SomeActivities) SomeOperation(ctx context.Context, input *some.SomeOperationInput) (*some.SomeOperationOutput, error) {
	client, err := a.getClient()
	if err != nil {
		return nil, err
	}

	return a.client.AddTagsToCertificateWithContext(ctx, input)
}
```

We could further optimize the creation of the client if we want: don't create the client in the constructor, but use a `sync.Once` to create it for the first time. Then anyone who uses the worker and doesn't want to use all the services doesn't have to pay the initialization and memory cost of initializing all clients. (Then again: client initialization is cheap)

**Comment by sagikazarmark** [comment]
@mfateev does it sound like an acceptable solution?

**Comment by sagikazarmark** [comment]
Fixed in #7
